src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +148,297 @@ public interface IFileWatchingToken
+                        // Watching operations cannot be combined
+                        break;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
+                        return true;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
+                        return true;
+
+                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
+                        var cookies = new List<uint>(_cookies);
+                        cookies.AddRange(other._cookies);
+                        combined = UnwatchDirectories(cookies);
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().Add(other._token));
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
+                        return true;
+
+                    default:
+                        break;
+                }
+
+                combined = default;
+                return false;
+            }
+
+            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            {
+                switch (_kind)
+                {
+                    case Kind.None:
+                        return;
+
+                    case Kind.WatchDirectory:
+                        var cookie = await service.AdviseDirChangeAsync(_directory, watchSubdirectories: true, _sink, cancellationToken).ConfigureAwait(false);
+                        _cookies.Add(cookie);
+
+                        if (_filter != null)
+                            await service.FilterDirectoryChangesAsync(cookie, new string[] { _filter }, cancellationToken).ConfigureAwait(false);
+
+                        return;
+
+                    case Kind.WatchFile:
+                        _token.Cookie = await service.AdviseFileChangeAsync(_directory, _fileChangeFlags, _sink, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchFile:
+                        await service.UnadviseFileChangeAsync(_token.Cookie!.Value, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchDirectories:
+                        Contract.ThrowIfFalse(_cookies is not null);
+                        foreach (var unwatchCookie in _cookies)
+                        {
+                            await service.UnadviseDirChangeAsync(unwatchCookie, cancellationToken).ConfigureAwait(false);
```

Comment:
i believe there was a convo about bulk unadvising?  if that is applicable, could you do in a followup pr?

Reply:
Yes, it requires APIs which are not present in main. Will submit as a follow-up to main-vs-deps (mentioned in first post).

Reply:
perfect.